### PR TITLE
support variable-length string type for index

### PR DESCRIPTION
**Pull request by bright-starry-sky**

### CODE REVIEW

File: src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp [hunk 1]
```diff
@@ -104,18 +107,35 @@ void CreateEdgeIndexProcessor::process(const cpp2::CreateEdgeIndexReq& req) {
...
+        cpp2::ColumnDef col = *iter;
+        if (col.type.get_type() == meta::cpp2::PropertyType::STRING) {
+            if (!field.__isset.type_length) {
+                LOG(ERROR) << "No type length set : " << field.get_name();
+                handleErrorCode(cpp2::ErrorCode::E_INVALID_PARM);
+                onFinished();
+                return;
+            }
+            col.type.set_type(meta::cpp2::PropertyType::FIXED_STRING);
+            col.type.set_type_length(*field.get_type_length());
```
Comment (panda-sheep) [inline src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp:131]: `STRING `   to    `FIXED_STRING`， 👍

**Review by panda-sheep — APPROVED** [review]

**Comment by bright-starry-sky** [comment]
rebased and resolved conflict

File: src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp [hunk 2]
```diff
@@ -99,18 +102,35 @@ void CreateEdgeIndexProcessor::process(const cpp2::CreateEdgeIndexReq& req) {
...
+        } else if (field.__isset.type_length) {
+            LOG(ERROR) << "No need to set type length : " << field.get_name();
+            handleErrorCode(cpp2::ErrorCode::E_INVALID_PARM);
+            onFinished();
+            return;
+        }
```
Comment (panda-sheep) [inline src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp:132]: If it is a non-string type, must an error be reported when type_length is set?

Comment (bright-starry-sky) [reply src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp:132]: > If it is a non-string type, must an error be reported when type_length is set?

yeah，if is an no-string type, don't need to set the length. else an error code will be sent.

**Review by panda-sheep — APPROVED** [review]
Excellent!

File: src/meta/test/MetaClientTest.cpp [hunk 3]
```diff
@@ -783,7 +821,12 @@ TEST(MetaClientTest, EdgeIndexTest) {
         ASSERT_EQ(Status::Error("not existed!"), result.status());
     }
     {
-        std::vector<std::string>&& fields {"edge_0_col_0",  "edge_0_col_0"};
+        std::vector<cpp2::IndexFieldDef> fields;
+        cpp2::IndexFieldDef field1, field2;
+        field1.set_name("edge_0_col_0");
```
Comment (darionyaphet) [inline src/meta/test/MetaClientTest.cpp:826]: index fields are the same ?

File: src/meta/test/ProcessorTest.cpp [hunk 4]
```diff
@@ -2265,8 +2266,9 @@ TEST(ProcessorTest, TagIndexTest) {
         cpp2::CreateTagIndexReq req;
         req.set_space_id(1);
         req.set_tag_name("tag_0");
-        std::vector<std::string> fields{"tag_0_col_0"};
-        req.set_fields(std::move(fields));
+        cpp2::IndexFieldDef field;
+        field.set_name("tag_0_col_0");
+        req.set_fields({field});
```
Comment (darionyaphet) [inline src/meta/test/ProcessorTest.cpp:2271]: why don't use move ?

File: src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp [hunk 5]
```diff
@@ -14,10 +14,13 @@ void CreateEdgeIndexProcessor::process(const cpp2::CreateEdgeIndexReq& req) {
     CHECK_SPACE_ID_AND_RETURN(space);
     const auto &indexName = req.get_index_name();
     auto &edgeName = req.get_edge_name();
-    auto &fieldNames = req.get_fields();
+    const auto &fields = req.get_fields();
 
-    std::set<std::string> columnSet(fieldNames.begin(), fieldNames.end());
-    if (fieldNames.size() != columnSet.size()) {
+    std::set<std::string> columnSet;
```
Comment (darionyaphet) [inline src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp:19]: any advantage?

Comment (bright-starry-sky) [reply src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp:19]: fields is not a string list.

File: src/meta/test/ProcessorTest.cpp (hunk 4 above)
Comment (bright-starry-sky) [reply src/meta/test/ProcessorTest.cpp:2271]: In this case, it works either way.Memory will be destroyed when this code block is complete.

**Comment by bright-starry-sky** [comment]
Close this PR. since #158 was merged.

**Pull request closed without merging**